Backends entrypoints

aurghs opened the pull request:
- It's an update of @jhamman  pull request https://github.com/pydata/xarray/pull/3166
- It uses `entrypoints`  module to detect the installed engines. The detection is done at `open_dataset` function call and it is cached. It raises a warning in case of conflicts.
- Add a class for the backend interface `BackendEtrypoint` instead of a function.

Modified files: 
- add plugins.py containing `detect_engines` function and `BackendEtrypoint`.
- dependencies file to add `entrypoints`. 
- backend.__init__ to add `detect_engines`
- apiv2.py and api.py do use `detect_engines`
- zarr.py, h5netcdf_.py, cfgrib.py to instatiate the `BackendEtrypoint`.

 - [x] Related to #3166
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`

pep8speaks commented:
Hello @aurghs! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`xarray/backends/__init__.py`](https://github.com/pydata/xarray/blob/d1849472d05abc55d37dc7de57246e1940f9ece1/xarray/backends/__init__.py):

> [Line 12:1](https://github.com/pydata/xarray/blob/d1849472d05abc55d37dc7de57246e1940f9ece1/xarray/backends/__init__.py#L12): [F401](https://duckduckgo.com/?q=pep8%20F401) '.plugins.detect_engines' imported but unused

jhamman commented on xarray/backends/plugins.py:
```diff
@@ -1,31 +1,89 @@
+        backend = backend.load()
+        engines[name] = backend
+
+        if backend.open_dataset_parameters is None:
+            open_dataset = backend.open_dataset
+            backend.open_dataset_parameters = detect_parameters(open_dataset)
```
I'm wondering if we can just require all backend entrypoints to provide the `open_dataset_parameters` attribute. It seems to me this introduces a fair bit of complexity that most backend developers wont utilize.

aurghs replied:
The idea was to avoid to duplicate the backend interface: in the function signature and in the `BackendEntrypoints. open_dataset_parameters`. If the backend needs to change the signature, then changes must be done in two places. The explicit declaration could be useful in case the backend prefers to use `**kwargs` or `*args`.
I agree we add a little bit of complexity, usually, I prefer to reduce complexity, but in this case, the backend interface seems to friendly.

jhamman commented on xarray/backends/plugins.py:
```diff
@@ -1,31 +1,89 @@
+import entrypoints  # type: ignore
+
+
+class BackendEntrypoint:
+    __slots__ = ("open_dataset", "open_dataset_parameters")
+
+    def __init__(self, open_dataset, open_dataset_parameters=None):
+        self.open_dataset = open_dataset
+        self.open_dataset_parameters = open_dataset_parameters
+
```
If we required `open_dataset_parameters` in the constructor, we could just use a `namedtuple` here. If we foresee any subclassing of the `BackendEntrypoint` class, then I suggest we stick with what you have here.

jhamman reviewed (commented):
Thanks @aurghs for pushing this forward. I have a few small comments but think this is otherwise quite close. We should discuss (tomorrow I suppose) a plan for testing the entrypoint discovery logic and error handling. 

Just to bring to the attention of the rest of @pydata/xarray, this does add a new, small dependency on the [`entrypoints`](https://entrypoints.readthedocs.io/en/latest/index.html) package.

aurghs commented:
- I have replaced entrypoints with pkg_resources. I can't see any drawback in this change, but only advantages, the main one is that we remove an external dependency.
- I have added the tests
- I didn't remove the signature inspection. Let me know if you both (@jhamman and @shoyer) agree to remove it.

keewis commented on xarray/tests/test_plugins.py:
```diff
@@ -0,0 +1,97 @@
+    assert len(entrypoints) == 2
+
+
+def test_remove_duplicates_wargnings(dummy_duplicated_entrypoints):
```
```suggestion
def test_remove_duplicates_warnings(dummy_duplicated_entrypoints):
```

keewis commented on xarray/tests/test_plugins.py:
```diff
@@ -0,0 +1,97 @@
+        "engine1 = xarray.tests.test_plugins:backend_1",
+        "engine2 = xarray.tests.test_plugins:backend_2",
+    ]
+    entrypoints = []
+    for spec in specs:
+        entrypoints.append(pkg_resources.EntryPoint.parse(spec))
```
same here, this could be transformed into a comprehension

keewis reviewed (commented):
I have a few suggestions which I feel will improve the readability, but I'm not sure if that's objectively true for each of these.

Also, as a note to myself and to others trying to test this (I spent way too much time trying to figure that out): after changing the entrypoints in `setup.cfg`, the package has to be reinstalled for the changes to be visible (editable installs only allow changing the code without reinstalling).

jhamman commented:
@aurghs - I think we're basically done here. Can you clean up the remaining failing tests and we'll get this merged?

alexamici commented:
@keewis WRT your comments on list comprehensions we at B-Open prefer to use them only when they make the code one line and actually prefer explicit loops otherwise.

This is both because we find them more readable as code and because tracebacks as well are easier to parse when they point to the statement that raised.

keewis commented:
no worries, as I said when posting these suggestions I do know that there are different opinions about comprehensions or generator expressions. I might have been more explicit about this, though, I meant to say that you may choose not to apply suggestions you do not agree with.
